### PR TITLE
Remove asserts from Kroxylicious

### LINKED ISSUE #45: Remove `assert` statements
Either truly enforce the invariant also at runtime, or don't. As assertions are not enabled in many production environments, they represent some odd go-between.

### PR DESCRIPTION
**Pull request by SamBarker**
Partial fix for #45 as I think it makes sense to introduce checkstyle and associated rules separately. 

I've generally sided with removing the asserts as they are on the hot-path for the proxy. However I kept the channel handler context checks as I think they are invoked less often and would point to genuine "wiring" issues, but I can't quite convince myself that the checks are cheep enough to keep enough on that execution path.   

Fixes #45

### CODE REVIEW

**Comment by gunnarmorling** [comment]
Thanks, @SamBarker!

Any thoughts on that one, @tombentley? I'm sympathetic towards it, as I always feel asserts are sort of a Schrödinger thing, halfway there, but then not. But you added them initially, so it's your call really.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyFrontendHandler.java [hunk 1]
```diff
@@ -140,7 +140,7 @@ public void channelRead(final ChannelHandlerContext ctx, Object msg) {
     }
 
     public void outboundWritabilityChanged(ChannelHandlerContext outboundCtx) {
-        assert this.outboundCtx == outboundCtx;
+        LOGGER.warn("Outbound Channel write ability change notification for inappropriate context. {} received notification for {}", this.outboundCtx, outboundCtx);
```
Comment (tombentley) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyFrontendHandler.java:143]: Should this be guarded by an `if (this.outboundCtx != outboundCtx)`? 
Should we use the single word 'writability'

File: kroxylicious/src/main/java/io/kroxylicious/proxy/frame/DecodedFrame.java [hunk 2]
```diff
@@ -81,7 +81,6 @@ public short apiVersion() {
     @Override
     public final int estimateEncodedSize() {
         if (headerAndBodyEncodedLength != -1) {
-            assert serializationCache != null;
```
Comment (tombentley) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/frame/DecodedFrame.java:84]: Can we have a comment about why it must be non-null at this point?

**Comment by tombentley** [comment]
I think @franz1981 added at least some of the uses of `assert`, so let's see if he has a strong opinion for keeping them.

**Comment by gunnarmorling** [comment]
That's fair. To expand a bit on my personal stance: as you never no whether asserts are enabled or not (not everyone runs `java -ea ...`, and they are not enabled by default), it's just not something you can rely on. So if for instance you'd want to make sure a method parameter passed by some caller adheres to some precondition, you'd still have to check and validate that parameter, so as to reliably raise a meaningful exception or return an error message. Some value I could see about having assertions is as means of documenting assumptions within internal code, but somehow that approach never really resonated with me.

**Comment by SamBarker** [comment]
The other argument I've seen for asserts in performance sensitive settings is that one enables them during test runs, even in integration or test environments as a way of sanity checking things in controlled environments but still achieving maximum performance in latency sensitive environment. Particularly used for things like ensuring thread consistency which are difficult to check otherwise.

I've never entirely bought into the argument, but I can see the validity in it.

That was my thinking around inflating some instances to full blown if checks and removing the others entirely.

**Comment by franz1981** [comment]
Agree with everyone here :)
I tends to use assert to protect code I have optimized ie which shape is not reflecting domain requirements
And assume specific preconditions I cannot check due to being hot path or to check concurrent invariant that I cannot and don't want to check at runtime.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyFrontendHandler.java (hunk 1 above)
Comment (SamBarker) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyFrontendHandler.java:143]: Yes, this certainly should be guarded and I'll go have a word with the IntelliJ spell checker.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/frame/DecodedFrame.java (hunk 2 above)
Comment (SamBarker) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/frame/DecodedFrame.java:84]: As in why it would be a problem if it was `null` ? Or why we shouldn't get here without it being `null`?

Comment (tombentley) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/frame/DecodedFrame.java:84]: The latter I think: It should never be null by the point this method gets called

**Comment by tombentley** [comment]
> The other argument I've seen for asserts in performance sensitive settings is that one enables them during test runs, even in integration or test environments as a way of sanity checking things in controlled environments but still achieving maximum performance in latency sensitive environment. Particularly used for things like ensuring thread consistency which are difficult to check otherwise.

Exactly. The problem with that argument is that you're at the mercy of your test coverage: Not doing those checks in production means you'll never learn if they can be violated in some edge case only found in some specific prod configuration/environment, which is precisely what you need to find out in order to improve the software. In that respect they're like seatbelts that break when you actually need them.

**Comment by gunnarmorling** [comment]
Hey @SamBarker, so we had a quick chat about this in the weekly Kroxylicious call. The consencus we achieved was this:

* Let's keep using asserts for documenting invariants in internal code (private methods, etc.) in a executable way; as tests are run with assertions enabled, this can help to spot bugs which would otherwise go unnoticed
* Let's not use asserts for validating parameters of public method invocations, as we cannot rely on assertions to be enabled at production runtime; rather check via if ... and throw an `IllegalArgumentException` in case of any violations

If that makes sense to you too, could you perhaps take a look at the existing asserts in the code base, and if they fall into the first category, keep them, and if they fall into the second category, rework them? It would also be nice to have that reasoning above documented in some kind of contribution guide (probably needs creation of a new document _CONTRIBUTING.md_. WDYT?

**Comment by SamBarker** [comment]
I've added `CONTRIBUTING.md` as part of https://github.com/kroxylicious/kproxy/pull/78 as that seemed like a more natural home.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/frame/DecodedFrame.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (SamBarker) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/frame/DecodedFrame.java:84]: bump. Does anyone have any thoughts on the best way forward here.  Expand the guard clause, keep the assert or delete?

**Comment by SamBarker** [comment]
Closing this as stale. Will need to revisit the question of asserts at some point.
